### PR TITLE
fix(runner): make the asCell scope cap mean the same thing on every route

### LINKED ISSUE #5230: asCell scope cap is enforced by resolveAsCell() but not by the value projection or a key() chain past the handle
## What

For one cell, one schema, one stored link, three routes to the same handle give two different answers about whether the `asCell` descriptor's scope cap applies.

Setup — the descriptor caps follows at `space`, the stored link is `session`, so resolution is meant to be blocked:

```ts
const outerSchema = {
  type: "object",
  properties: {
    handle: { ...innerSchema, asCell: [{ kind: "cell", scope: "space" }] },
  },
  required: ["handle"],
};
const inner = runtime.getCell(space, "inner", innerSchema, tx, "session"); // session-scoped
inner.set({ field: "secret" });
outer.set({ handle: inner });
```

Measured on `8a0e0d4ac`:

| route | result | cap |
| --- | --- | --- |
| `outer.key("handle").get().get()` | `{"field":"secret"}` | **ignored** |
| `outer.key("handle","field").get()` | `"secret"` | **ignored** |
| `outer.key("handle").resolveAsCell()` | resolves to the `Undefined@1` data link | honored |

The schema cap is visible where you'd expect and then disappears one hop later:

```
getSchemaScopeCap(outer.key("handle").schema)         -> "space"
getSchemaScopeCap(outer.key("handle","field").schema) -> undefined
```

## Why it happens

Two independent gaps:

1. **`key()` past the handle drops the descriptor.** `key()` narrows via `getSchemaAtPath`, which walks the asCell schema's `properties` and returns the leaf schema (`{"type":"string"}`) without the ancestor's `asCell` entry. The cell's link then carries only that leaf schema, so `schemaScopeForLink()` in `link-resolution.ts` finds no cap and `canFollowLinkHop()` permits the session-scoped follow. `key()`'s own comment says scope "lives in the schema (top-level and asCell entries) and is resolved later as a follow cap during reads" — that holds for one hop, not two.

2. **The value projection doesn't consult the cap at all.** `resolveAsCell()` goes through `resolveLink()` and gets the cap; the `Cell` that `validateAndTransform` hands back for an asCell slot does not, so reading through it bypasses the check even though the cap is still present on that cell's schema.

## Why it matters

Whether a capped handle is enforced currently depends on which materialization a caller happens to use, and callers pick between them for unrelated reasons. `PiecePropIo.get` is the live example: it blocks `["handle"]` (routes through `resolveAsCell()`) and leaks `["handle","field"]` (a `key()` chain) — this is pre-existing on `main`, not introduced by #5222 or #5229. A pathless root read is a third answer again: it returns the raw sigil link, handing the caller the blocked target's id, space, and scope.

Any code navigating a path across an asCell boundary with `key()` inherits gap 1, so this is not piece-specific.

## Where the fix belongs

In the runner, not at the call sites. Roughly: carry the ancestor's cap forward when `key()` narrows past an asCell (either preserved on the child schema or tracked on the link beside it, so `schemaScopeForLink` still sees it), and make the asCell value projection apply the same check `resolveAsCell()` does. Point-fixing consumers by routing everything through `resolveAsCell()` would paper over gap 1 while leaving every other `key()` user exposed.

## Notes

Found while reviewing #5229. Codex flagged an adjacent claim there — that dropping the hand-rolled per-segment traversal regressed the path-through case — which measurement did not support: `["handle","field"]` returns `"secret"` on `main` too. It did surface this, which is the real problem underneath. #5229 keeps the `resolveAsCell()` route for the terminal case so it stays behavior-identical to `main` rather than encoding either behavior as intended.

### PR DESCRIPTION
**Pull request by seefeldb**
## Why

Closes #5230.

An `asCell` entry's `scope` is a follow cap — it bounds which link scopes a read arriving through that schema may follow. It is what lets a space-scoped list require that every handle in it points somewhere every reader in the space can resolve (`packages/runner/test/pattern-scope.test.ts` builds exactly that: `authorProfile` capped at `space` inside a space-scoped `messages` array).

Only one of the three routes to a handle enforced it. For one cell, one schema, one stored link:

```
cell.key("handle").get().get()      -> {"field":"secret"}   cap ignored
cell.key("handle", "field").get()   -> "secret"             cap ignored
cell.key("handle").resolveAsCell()  -> Undefined@1          cap honored
```

So whether a capped handle held depended on which materialization a caller happened to reach for, and callers pick between them for reasons that have nothing to do with scope. `PiecePropIo.get` is the live example: it blocks `["handle"]` and leaks `["handle","field"]`, purely because those two paths route differently.

## What changed

Two independent gaps.

**1. `key()` past the handle dropped the declaring schema.** Narrowing goes through `getSchemaAtPath`, which walks the `asCell` schema's `properties` and returns the leaf (`{"type":"string"}`), so `resolveLink`'s `schemaScopeForLink` had nothing to check one hop later:

```
getSchemaScopeCap(outer.key("handle").schema)         -> "space"
getSchemaScopeCap(outer.key("handle","field").schema) -> undefined
```

`key()` now records a cap it narrows past on the link (`scopeCaps: [{depth, scope}]`), and `resolveLink` consults the cap declared at the depth where it actually *finds* the hop rather than the leaf's. `scopeCaps` is read-side only: never serialized into a sigil link (`createSigilLinkFromParsedLink` builds field by field), never part of link identity (`areNormalizedLinksSame` compares id/space/scope/path), and never stamped onto a followed link's own scope — CT-1623 stays intact.

**2. The value projection never consulted the cap at all.** An `asCell` boundary is the one place a link is turned into a handle *instead of* being followed, so `followPointer`'s check never runs for it. Both boundary sites — the traverser's `getNextCellLink` and `validateAndTransform`'s top-level `asCell` path — now apply the same rule and resolve to undefined-data. `undefinedDataLink` is exported from `link-resolution.ts` so all three routes produce the same shape by construction.

## Test plan

New `packages/runner/test/ascell-scope-cap.test.ts` pins all three routes to one answer across four schema shapes. The three permissive cases (cap admits the link, no cap declared, link no narrower than the cap) already passed before the fix and are here as regression guards — the risk in this change is over-blocking, not under-blocking. A fifth test pins the `scopeCaps` plumbing directly.

- `deno fmt` / `deno lint` / `deno check` on all changed files — clean
- **`packages/runner` — 1103 passed / 5752 steps / 0 failed**, via the package's own task (`--preload=test/clock-preload.ts`; without the clock preload the suite has 108 pre-existing timing failures on unmodified `main` too, so that run is not a baseline)
- `packages/piece` — 31 tests / 368 steps
- `packages/patterns` — 66 tests / 32 steps
- `packages/fuse` — 347 tests
- `packages/cli` piece tests — 72 steps

## Follow-up

This unblocks removing the `resolveAsCell` routing that #5229 had to keep in `PiecePropIo.get`: with the cap enforced on the projection, the plain `key()` + projection read blocks a capped handle on its own. Verified against this branch — `["handle"]` returns `undefined` and `["handle","field"]` is blocked. One wrinkle to settle there: the blocked through-path currently falls through to the root read and surfaces as `Cannot access path "handle/field" - property "field" not found` rather than `undefined`, so the piece layer reports a blocked read as a missing one. Worth deciding deliberately rather than inheriting.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Enforces `asCell` scope caps on every route (projection, `key()` through-paths, deref, whole-object/array traversal) so capped handles behave the same everywhere and resolve to a unified undefined-data shape when blocked (fixes #5230). Also handles caps inside `anyOf`/`oneOf` and corrects cap rebasing across already-followed hops.

- Bug Fixes
  - Track caps via `scopeCaps`: seed from the starting address, correctly rebase across followed hops (fixed shift off-by-depth), keep the narrower cap when re-declared via `asSchema`, and preserve across `asSchema`. Read-only: not serialized, not part of identity, never stamped onto followed links.
  - Resolve using the cap at the hop’s depth: combine the asCell follow cap (including caps wrapped in `anyOf`/`oneOf`, narrowest wins) with the leaf’s; for ancestor hops use the narrower of the recorded cap and the leaf cap, keeping the leaf cap as a floor when no recorded caps exist.
  - Enforce the same cap check at all asCell boundaries at a single handle-mint site (projection, whole-object reads, arrays, redirects). Consult only the entry’s own asCell scope (ignore `schema.scope`), return `undefinedDataLink` (which drops `scopeCaps`) and skip deref logging when blocked; writes through blocked handles are refused.
  - Add tests for arrays, `anyOf`/`oneOf` compounds (null branches ignored; narrowest cap wins), cross-hop rebasing at non-root depth, positional caps, `asSchema` tightening, leaf-cap floor, no invented cap from `scope`, blocked writes, and all access routes.
  - Document the monotonic-cap rule on `scopeCaps`: caps persist across `asSchema` (cannot be lifted), with rationale added inline.

<sup>Written for commit 57a0ba79d23d33bde2e3556c17e92e7e79a2a612. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/commontoolsinc/labs/pull/5231?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by seefeldb** [comment]
Addressed the bot review plus a self-review pass (`claude-opus-5`, max effort). The self-review found **four real bugs**, two of which meant the cap still did not hold — including the change's own motivating shape.

## Bot review

- **cubic P1, root cap lost** — real. `key()` now seeds from the address it starts at. My first test for this was **vacuous** (an earlier `key()` had already recorded the cap); the revert-check caught it and the replacement fails without the seed.
- **cubic P1, leaf-cap fallback dropped for ancestor hops** — real. The leaf cap is kept as a *floor* (narrower of recorded and leaf) rather than replaced, so this change can only ever block more than before, never less.
- **cubic P2, blocked-handle identity** — done. One helper, `cappedHandleLink`, builds every blocked handle from the source address.
- **codex P2, `asSchema` keeps stale caps** — partly right, and more interesting than stated. Caps deliberately survive `asSchema` (dropping them makes it a cap bypass, and `cellWithScopedLinkRequiredsRelaxed` is an `asSchema` call that must carry them). But the dedup guard let a *looser* recorded cap shadow a *tighter* one re-declared at the same depth — so `asSchema` could not lift a cap and could not apply one either. It now keeps the narrower.
- **Coverage gate** — correct and load-bearing: the traverser boundary had no test at all. See below.

## Self-review findings

1. **Array-element handles bypassed the check entirely.** By the time an element reaches the boundary, traversal has already stepped past its link, so the sigil branch never ran and the handle was built onto the target. This is exactly the shape the change exists for. The check moved to one choke point applied to the finished handle link at all four boundary sites.
2. **A capped handle below an already-followed link went unchecked.** Caps were dropped at every hop, so a cell whose root value is a link — ordinary for pattern argument/result cells — lost them first. Caps for the remaining path are now rebased onto the hop target.
3. **`scope` beside an uncapped `asCell` was read as a follow cap**, blocking reads `main` allows. Boundary checks now consult only the `asCell` entry's own scope.
4. **The leaf-cap floor is not unreachable.** My comment said it was and was wrong — `getCellFromLink` with a multi-segment path carries no recorded caps, and dropping the floor follows a link `main` blocks. Now pinned by a test.

Verified clean by the same pass: no writes through a blocked handle (`ReadOnlyAddressError`), reactivity intact (a subscriber re-fires with the real value when the link is repointed to a permitted scope), `scopeCaps` never reaches a sigil / `getAsLink` / `toJSON` / link identity / any cache key, and streams and `user`-capped handles unaffected.

## Known gap, not fixed here

An `asCell` cap wrapped in `anyOf`/`oneOf` is not enforced on any route — `getSchemaScopeCap` only inspects the top level. Pre-existing on `main`, unchanged by this PR, and a real shape in this repo (`consumeAsCellProjectionSchema` handles it). Filed separately rather than widened into this PR.

## Validation

- `deno fmt` / `deno lint` / `deno check` — clean
- **`packages/runner` — 1104 passed / 5761 steps / 0 failed** (clock preload)
- `packages/piece` 368 steps · `packages/patterns` 66 tests · `packages/fuse` 347 tests · `packages/cli` piece 72 steps
- 14 tests in `ascell-scope-cap.test.ts`; **every guard revert-verified** — cross-hop rebase, leaf floor, dedup narrowing, and the boundary check each fail a test when removed.

**Comment by seefeldb** [comment]
Both outstanding items fixed, plus the `anyOf` gap I had deferred.

**Rebase off-by-depth (P1) — real.** A cap at old depth `k` lands at `(target base length) + (k - depth)`, so the shift is target-path length minus ours, not minus the *remaining* length; mine was off by exactly the hop's depth. My test for the rebase used a link at the **root**, where depth is 0 and both formulas agree — which is why it hid. The new test puts the ancestor link at `wrap` and the cap at `wrap/handle`.

**Compound `asCell` cap (P2) — real, and worse than the comment said.** It was enforced on *no* route. Two reasons: `getSchemaScopeCap` only reads the top level, and a **mixed** compound (`[<capped handle>, {type:"null"}]`) never mints a handle at all — `hasAsCell` is false for it and `asCellCompoundSchemaForValue` does not match a stored link, so the link is simply followed. So this had to be caught at the resolver, not only where handles are built.

`getAsCellFollowScopeCap` looks through `anyOf`/`oneOf`, skips branches declaring no `asCell` (a `null` alternative is not a handle), and takes the narrowest of the rest. It is consulted by the resolver's cap lookup, by `key()`'s recording, and where handles are minted. It deliberately drops `getSchemaScopeCap`'s `schema.scope` fallback — authors write `scope` to mean "this value lives at that scope", and reading it as a follow cap invents a restriction nobody asked for.

**Simplification that fell out.** Chasing the `anyOf` case turned up the actual choke point: `TransformObjectCreator.createObject` is the single place a handle is minted, and every route funnels through it. The check lives there now, and the four boundary wrappers I had added to `traverse.ts` are gone.

**Also pinned: why `scopeCaps` carries a depth** rather than collapsing to one cap. A `user`-capped container holding a `space`-capped handle — the shape `pattern-scope.test.ts` builds — needs each cap to govern the hop at its own position. Collapsing to the narrowest would block the container's own legitimate user-scoped link. New test asserts the wide hop still resolves while the narrow one is enforced.

## Validation

- `deno fmt` / `deno lint` / `deno check` — clean
- **`packages/runner` — 1105 passed / 5764 steps / 0 failed** (clock preload)
- `packages/piece` 368 steps · `packages/patterns` 66 tests · `packages/fuse` 347 tests · `packages/cli` piece 72 steps
- 17 tests in `ascell-scope-cap.test.ts`; all four new guards revert-verified (rebase shift, compound leaf cap, `createObject` check, `key()` compound recording) — each fails a test when removed.

One process note: an intermediate full-suite run went red on my own `anyOf` test, and the cause was me — a revert-check used `git checkout --` on a file with uncommitted work and wiped the edit it was supposed to restore. Caught by running the full suite rather than the focused file.

File: packages/runner/src/cell.ts [hunk 1]
```diff
@@ -1931,10 +1931,25 @@ export class CellImpl<T extends FabricValue>
...
+          depth: path.length,
+          scope: cap,
+        }];
+
       currentLink = {
         ...currentLink,
-        path: [...currentLink.path, key.toString()] as string[],
+        path,
         schema: childSchema,
+        ...(scopeCaps !== undefined && { scopeCaps }),
```
[the remark this answers is not in the document]
Comment (seefeldb) [reply packages/runner/src/cell.ts:1974]: Seen, and deliberately not done — but you're right that the asymmetry is real, and it deserved an answer here rather than only in the PR thread.

**Why caps are kept.** `cellWithScopedLinkRequiredsRelaxed` — the piece read boundary itself — ends in `cell.asSchema(relaxed)` (`packages/runner/src/piece-helpers.ts:285`). A narrowed cell's *ancestor* caps live only in `scopeCaps`; they are not recoverable from the leaf schema, because narrowing is what destroyed the declaring schema in the first place. Clearing on `asSchema` would therefore make the read boundary silently lift the caps it exists to enforce, and the same would apply to any other `asSchema`-based read helper. That is a worse failure than the surprise you describe.

**What did change.** Your comment surfaced a genuine bug next to the one you named. The dedup guard skipped on depth alone, so a *tighter* cap re-declared at an already-recorded depth was dropped — meaning `asSchema` could neither lift a cap nor apply one. It now keeps the narrower, so the rule is monotonic: **reinterpreting an address can tighten a cap, never lift one.** Pinned by `lets asSchema tighten a cap it cannot loosen`.

The construction-history asymmetry you point at is inherent to carrying provenance on the link, and it is the safe direction of that trade: the path that "remembers more" is the one that blocks. I've written the rule and its rationale onto the `scopeCaps` declaration (`link-types.ts`) so the next reader doesn't have to re-derive it from a test.
